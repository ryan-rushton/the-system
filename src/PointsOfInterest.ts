import { createRef } from 'react';

/**
 * All the points of interest in the solar system that get rendered.
 * Values are obtained from either https://nssdc.gsfc.nasa.gov/planetary/factsheet/ or wikipedia.
 *
 * The following properties may or may not exist on each point of interest,
 * - id: string - A unique string used for data-testid identifiers for tests.
 * - ref: RefObject<HTMLDivElement> - A ref so that we can scroll to the object when it is selected in the nav menu.
 * - display: string - The name of the point.
 * - radius: number - The radius of the point of interest (includes rings if visible), units are km.
 * - distance: number - The points distance from the sun, units are km.
 * - orbitalPeriod: number - How many days it takes the point to orbit the sun.
 * - Satellites: PointOfInterest[] - An array of satellites that orbit the point. These have radius, distance,
 *  orbitalPeriod just like the points.
 */
export const pointsOfInterest = {
  sun: {
    id: 'sun',
    ref: createRef<HTMLDivElement>(),
    radius: 695508,
    distance: 0,
    orbitalPeriod: 1,
  },
  mercury: {
    id: 'mercury',
    ref: createRef<HTMLDivElement>(),
    radius: 2440,
    distance: 57900000,
    orbitalPeriod: 88,
  },
  venus: {
    id: 'venus',
    ref: createRef<HTMLDivElement>(),
    radius: 6052,
    distance: 108200000,
    orbitalPeriod: 224.7,
  },
  earth: {
    id: 'earth',
    ref: createRef<HTMLDivElement>(),
    radius: 6378,
    distance: 149600000,
    orbitalPeriod: 365.2,
    satellites: [
      {
        id: 'moon',
        distance: 384400,
        orbitalPeriod: 27.3,
        radius: 1737,
      },
    ],
  },
  mars: {
    id: 'mars',
    ref: createRef<HTMLDivElement>(),
    radius: 3396,
    distance: 227900000,
    orbitalPeriod: 687,
  },
  theBelt: {
    id: 'theBelt',
    ref: createRef<HTMLDivElement>(),
  },
  jupiter: {
    id: 'jupiter',
    ref: createRef<HTMLDivElement>(),
    radius: 71498,
    distance: 778600000,
    orbitalPeriod: 4331,
    satellites: [
      {
        id: 'io',
        distance: 421700,
        orbitalPeriod: 1.77,
        radius: 1822,
      },
      {
        id: 'europa',
        distance: 670900,
        orbitalPeriod: 3.55,
        radius: 1561,
      },
      {
        id: 'ganymede',
        distance: 1070400,
        orbitalPeriod: 7.15,
        radius: 2634,
      },
      {
        id: 'callisto',
        distance: 1882700,
        orbitalPeriod: 16.69,
        radius: 2410,
      },
    ],
  },
  saturn: {
    id: 'saturn',
    ref: createRef<HTMLDivElement>(),
    radius: 80000 + 58232, // making the ring radius clear
    distance: 1433500000,
    orbitalPeriod: 10747,
    satellites: [
      {
        id: 'titan',
        distance: 1221870,
        orbitalPeriod: 15.95,
        radius: 2574,
      },
      {
        id: 'rhea',
        distance: 527108,
        orbitalPeriod: 1.77,
        radius: 764,
      },
    ],
  },
  uranus: {
    id: 'uranus',
    ref: createRef<HTMLDivElement>(),
    radius: 25559,
    distance: 2872500000,
    orbitalPeriod: 30589,
  },
  neptune: {
    id: 'neptune',
    ref: createRef<HTMLDivElement>(),
    radius: 24764,
    distance: 4495100000,
    orbitalPeriod: 59800,
  },
  pluto: {
    id: 'pluto',
    ref: createRef<HTMLDivElement>(),
    radius: 1185,
    distance: 5906400000,
    orbitalPeriod: 90560,
  },
} as const;
