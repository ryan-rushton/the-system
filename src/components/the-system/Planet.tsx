import { useContext, type RefObject } from 'react';
import { pointsOfInterest } from '../../PointsOfInterest';
import { AppContext } from '../../context/SystemContext';
import styles from './Planet.module.scss';
import { CelestialBody, type CelestialBodyProps } from './celestial-body/CelestialBody';

/** Represents details used for the moon of a planet. */
interface Satellite {
  /** A value to use on data-testid for cypress tests. Likely the display from pointsOfInterest. */
  id: string;
  /** The radius in km */
  radius: number;
  /** The distance from what it is orbiting in km */
  distance: number;
  /** Its orbital period in days. */
  orbitalPeriod: number;
}

/**
 * A planet in the solar system. This applies multipliers to values before they get passed to CelestialBody.
 */
export function Planet({
  id,
  className,
  distance,
  radius,
  orbitalPeriod,
  scrollToRef,
  satellites,
}: {
  /** A value to use on data-testid for cypress tests. Likely the display from pointsOfInterest. */
  id: string;
  /** Class name to give the element, passed to CelestialBody. */
  className: string;
  /** A ref so that we can scroll to planet when it is selected in the nav menu. */
  scrollToRef: RefObject<HTMLDivElement>;
  /** The radius of the planet in km's. */
  radius: number;
  /** The distance from the planet to the Sun. */
  distance: number;
  /** The number of days the planet takes to orbit the sun */
  orbitalPeriod: number;
  /** Things that orbit the planet, all moons at this stage. */
  satellites?: readonly Satellite[];
}) {
  const {
    systemRadius,
    multipliers: { distanceMultiplier, sizeMultiplier, orbitalPeriodMultiplier, satelliteDist },
  } = useContext(AppContext);

  const adjustedDistance = distance * distanceMultiplier + (pointsOfInterest.sun.radius - radius) * sizeMultiplier;
  const adjustedRadius = radius * sizeMultiplier;
  const adjustedOrbitalPeriod = orbitalPeriod * orbitalPeriodMultiplier;

  const satelliteToCelestialBody = (satellite: Satellite): CelestialBodyProps => {
    // Make sure the moons are at least 1px so they can be seen.
    const satelliteRadius = Math.max(1, satellite.radius * sizeMultiplier);

    return {
      id: satellite.id,
      className: styles.satellite,
      distance: satellite.distance * satelliteDist + adjustedRadius - satelliteRadius,
      hasOrbitLine: true,
      orbitalPeriod: satellite.orbitalPeriod * orbitalPeriodMultiplier,
      radius: satelliteRadius,
      referenceRadius: adjustedRadius,
    };
  };

  return (
    <CelestialBody
      id={id}
      className={className}
      distance={adjustedDistance}
      hasOrbitLine={true}
      orbitalPeriod={adjustedOrbitalPeriod}
      radius={adjustedRadius}
      referenceRadius={systemRadius}
      satellites={satellites?.map(satelliteToCelestialBody)}
      scrollToRef={scrollToRef}
    />
  );
}
