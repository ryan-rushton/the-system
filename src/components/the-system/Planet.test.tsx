import { cleanup, render, screen } from '@testing-library/react';
import React from 'react';
import { pointsOfInterest } from '../../PointsOfInterest';
import Planet from './Planet';

describe('Planet', () => {
  afterEach(cleanup);

  test('it matches the snapshot', () => {
    const { jupiter } = pointsOfInterest;
    render(
      <Planet
        id={jupiter.id}
        className={'some-styles'}
        distance={jupiter.distance}
        orbitalPeriod={jupiter.orbitalPeriod}
        radius={jupiter.radius}
        satellites={jupiter.satellites}
        scrollToRef={jupiter.ref}
      />
    );

    expect(screen.getByTestId('jupiter-orbit-line')).toMatchSnapshot();
  });

  test('it renders the correct number of celestial bodies', () => {
    const { jupiter } = pointsOfInterest;
    render(
      <Planet
        id={jupiter.id}
        className={'some-styles'}
        distance={jupiter.distance}
        orbitalPeriod={jupiter.orbitalPeriod}
        radius={jupiter.radius}
        satellites={jupiter.satellites}
        scrollToRef={jupiter.ref}
      />
    );
    expect(screen.getByTestId('jupiter')).toBeInTheDocument();
    for (const satellite of jupiter.satellites) {
      expect(screen.getByTestId(satellite.id)).toBeInTheDocument();
    }
  });
});
