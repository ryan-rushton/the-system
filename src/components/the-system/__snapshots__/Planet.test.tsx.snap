// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Planet > it matches the snapshot 1`] = `
<div
  class="_orbit_ff166d orbitLine"
  data-testid="jupiter-orbit-line"
  style="animation: orbit 4331s linear infinite; height: 16267.508000000002px; left: 75888.8301142454px; top: 75888.8301142454px; width: 16267.508000000002px;"
>
  <div
    class="_celestialBody_ff166d some-styles"
    data-testid="jupiter"
    style="animation: planet-rotation 4331s linear infinite; height: 71.498px; left: -35.749px; top: 50%; width: 71.498px;"
    title="some-styles"
  >
    <div
      class="_orbit_ff166d orbitLine"
      data-testid="io-orbit-line"
      style="animation: orbit 1.77s linear infinite; height: 155.83800000000002px; left: -42.17000000000001px; top: -42.17000000000001px; width: 155.83800000000002px;"
    >
      <div
        class="_celestialBody_ff166d _satellite_aa3876"
        data-testid="io"
        style="animation: planet-rotation 1.77s linear infinite; height: 2px; left: -1px; top: 50%; width: 2px;"
        title="_satellite_aa3876"
      />
    </div>
    <div
      class="_orbit_ff166d orbitLine"
      data-testid="europa-orbit-line"
      style="animation: orbit 3.55s linear infinite; height: 205.678px; left: -67.09px; top: -67.09px; width: 205.678px;"
    >
      <div
        class="_celestialBody_ff166d _satellite_aa3876"
        data-testid="europa"
        style="animation: planet-rotation 3.55s linear infinite; height: 2px; left: -1px; top: 50%; width: 2px;"
        title="_satellite_aa3876"
      />
    </div>
    <div
      class="_orbit_ff166d orbitLine"
      data-testid="ganymede-orbit-line"
      style="animation: orbit 7.15s linear infinite; height: 285.57800000000003px; left: -107.04px; top: -107.04px; width: 285.57800000000003px;"
    >
      <div
        class="_celestialBody_ff166d _satellite_aa3876"
        data-testid="ganymede"
        style="animation: planet-rotation 7.15s linear infinite; height: 2.634px; left: -1.317px; top: 50%; width: 2.634px;"
        title="_satellite_aa3876"
      />
    </div>
    <div
      class="_orbit_ff166d orbitLine"
      data-testid="callisto-orbit-line"
      style="animation: orbit 16.69s linear infinite; height: 448.038px; left: -188.27px; top: -188.27px; width: 448.038px;"
    >
      <div
        class="_celestialBody_ff166d _satellite_aa3876"
        data-testid="callisto"
        style="animation: planet-rotation 16.69s linear infinite; height: 2.41px; left: -1.205px; top: 50%; width: 2.41px;"
        title="_satellite_aa3876"
      />
    </div>
  </div>
</div>
`;
